store/sql_preference_store.go

```diff
@@ -0,0 +1,226 @@
+	*SqlStore
+}
+
+func NewSqlPreferenceStore(sqlStore *SqlStore) PreferenceStore {
+	s := &SqlPreferenceStore{sqlStore}
+
+	for _, db := range sqlStore.GetAllConns() {
+		table := db.AddTableWithName(model.Preference{}, "Preferences").SetKeys(false, "UserId", "Category", "Name", "AltId")
+		table.ColMap("UserId").SetMaxSize(26)
+		table.ColMap("Category").SetMaxSize(32)
+		table.ColMap("Name").SetMaxSize(32)
+		table.ColMap("AltId").SetMaxSize(26)
+		table.ColMap("Value").SetMaxSize(128)
+	}
+
+	return s
+}
+
+func (s SqlPreferenceStore) UpgradeSchemaIfNeeded() {
+}
+
+func (s SqlPreferenceStore) CreateIndexesIfNotExists() {
+	s.CreateIndexIfNotExists("idx_preferences_user_id", "Preferences", "UserId")
+	s.CreateIndexIfNotExists("idx_preferences_category", "Preferences", "Category")
+	s.CreateIndexIfNotExists("idx_preferences_name", "Preferences", "Name")
+}
+
+func (s SqlPreferenceStore) Save(preference *model.Preference) StoreChannel {
+	storeChannel := make(StoreChannel)
+
+	go func() {
+		storeChannel <- s.save(s.GetMaster(), preference)
+		close(storeChannel)
+	}()
+
+	return storeChannel
+}
+
+func (s SqlPreferenceStore) save(queryable Queryable, preference *model.Preference) StoreResult {
+	result := StoreResult{}
+
+	if result.Err = preference.IsValid(); result.Err != nil {
+		return result
+	}
+
+	if err := queryable.Insert(preference); err != nil {
+		if IsUniqueConstraintError(err.Error(), "UserId", "preferences_pkey") {
+			result.Err = model.NewAppError("SqlPreferenceStore.Save", "A preference with that user id, category, name, and alt id already exists",
+				"user_id="+preference.UserId+", category="+preference.Category+", name="+preference.Name+", alt_id="+preference.AltId+", "+err.Error())
+		} else {
+			result.Err = model.NewAppError("SqlPreferenceStore.Save", "We couldn't save the preference",
+				"user_id="+preference.UserId+", category="+preference.Category+", name="+preference.Name+", alt_id="+preference.AltId+", "+err.Error())
+		}
+	} else {
+		result.Data = preference
+	}
+
+	return result
+}
+
+func (s SqlPreferenceStore) Update(preference *model.Preference) StoreChannel {
+	storeChannel := make(StoreChannel)
+
+	go func() {
+		storeChannel <- s.update(s.GetMaster(), preference)
+		close(storeChannel)
+	}()
+
+	return storeChannel
+}
+
+func (s SqlPreferenceStore) update(queryable Queryable, preference *model.Preference) StoreResult {
+	result := StoreResult{}
+
+	if result.Err = preference.IsValid(); result.Err != nil {
+		return result
+	}
+
+	if count, err := queryable.Update(preference); err != nil {
+		result.Err = model.NewAppError("SqlPreferenceStore.Update", "We couldn't update the preference",
+			"user_id="+preference.UserId+", category="+preference.Category+", name="+preference.Name+", alt_id="+preference.AltId+", "+err.Error())
+	} else {
+		result.Data = count
+	}
+
+	return result
+}
+
+func (s SqlPreferenceStore) SaveOrUpdate(preferences ...*model.Preference) StoreChannel {
```

Comment:
I can remove those. Now that I think of it, I think SqlChannelStore.Save might also do an upsert as well, but it can do it the trial and error way since it isn't wrapped in a transaction at all